### PR TITLE
[Event Hubs Client - Track Two Preview 1] Event Sender

### PR DESCRIPTION
**Pull request by jsquire**
# Summary

The intent of these changes is to complete the end-to-end infrastructure for sending events via the `EventSender`, from the API surface through the compatibility shim for the track one implementation.  Client operations have been validated at the unit level and basic sending scenarios have been smoke tested through Live tests.  

# Goals

- Ensure that an `EventSender` creates a transport-aware sender for containment to which it delegates operations.  This is based on an abstraction to isolate transport-level concerns from the core class, allowing different protocols, transports, compatibility shims, and test infrastructure to be used for delegation.

- Broker the compatibility between the track two API and the track one `EventDataSender` implementation, mapping types as needed.  Chief among them is mapping of the `EventData` model between the track one and track two incarnations.

- Allow for creation of an `EventSender` that is capable of interacting with the Event Hubs service for sending of events to either a specific partition or to the general Event Hub gateway for automatic partition routing.

- Review and revise the public surface area of the implementation, reducing scope as possible by internalizing constructs intended only to support project-level testing.

# Non-Goals

- Providing the implementation for the compatibility layer between the track two and track one `EventReceiver`; this will happen as part of future efforts.

- Supporting validation or other authorization-related tasks for non-SAS credential types; this will happen as part of future efforts.

- Cover the identified set of end-to-end live tests needed to fully exercise the target scenarios; this will happen as part of future efforts.

- Inclusion of the full set of supporting assets, such as README, samples, and ARM templates; these will happen as part of future efforts.

- Enabling test parallelism; this will be a stretch goal, but unlikely to occur in the time frame for the first track two preview.

# Details       

### General

- Ongoing review and updates to member visibility to reduce public scope and limit protected members.

- Shifting to a reflection-based approach over `internal` members for areas where the testing is focused in a tight scope.

###  Track One Token Provider

- Allowed for sub-paths of an event hub instance to be considered valid, which is needed for support of partition-based senders and receivers.

###  Event Sender

- Fleshed out the prototype skeleton to complete functionality using an abstraction for the transport-level concerns, making use of a compatibility shim to bridge to the track one `EventDataSender`.

###   Event Data Sender (Track One)

- Moved diagnostics and logging functionality into the core `EventDataSender` and out of the consuming classes (`PartitionSender` and `EventHubClient`) to remove duplication and allow the track two implementation to make use of the functionality implicitly.

###  Live Tests

- Added basic scenarios around the sending of events to smoke test the end-to-end scenarios; these will be extended with more complete coverage in a future workstream.

# Last Upstream Rebase

Tuesday, June 11, 2019  3:08pm (EDT)

# Resources

- [.NET Event Hubs Client: Track Two Proposal (First Preview)](https://gist.github.com/jsquire/75b3a3090b6b4553aa8ceb798b6fc87d)
- [.NET Event Hubs Client: Deferred Features and Functionality](https://gist.github.com/jsquire/f88922faa0f457b503234c6f168e20c9)

# Related and Follow-Up Issues

- [[Epic] Release Event Hubs Client Library Track Two Preview 1](https://github.com/Azure/azure-sdk-for-net/issues/6030) (#6030)

### CODE REVIEW

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs [hunk 1]
```diff
@@ -208,6 +208,37 @@ public override async Task<EventHubProperties> GetPropertiesAsync(CancellationTo
...
+        ///
+        /// <param name="senderOptions">The set of options to apply when creating the sender.</param>
+        ///
+        /// <returns>An event sender configured in the requested manner.</returns>
+        ///
+        public override EventSender CreateSender(EventSenderOptions senderOptions)
+        {
+           (TimeSpan minBackoff, TimeSpan maxBackoff, int maxRetries) = ((ExponentialRetry)senderOptions.Retry).GetProperties();
+
+           TrackOne.EventDataSender createSender()
```
Comment (AlexGhiondea) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs]: suggestion: rename this to `createSenderFactory`?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneSharedAccessSignatureTokenProvider.cs [hunk 2]
```diff
@@ -56,7 +56,11 @@ public TrackOneSharedAccessTokenProvider(SharedAccessSignature sharedAccessSigna
             Guard.ArgumentNotNullOrEmpty(nameof(resource), resource);
             Guard.ArgumentNotNegative(nameof(tokenValidityDuration), tokenValidityDuration);
 
-            if (!String.Equals(resource, SharedAccessSignature.Resource, StringComparison.InvariantCultureIgnoreCase))
+            // The requested resource may be a subpath of the current token resource, if a specific partition or other
+            // child object is referenced.  In that caes, the resource of the token should correspond to the starting
+            // segment of the requested resource.
+
+            if (resource.IndexOf(SharedAccessSignature.Resource, StringComparison.InvariantCultureIgnoreCase) != 0)
```
Comment (AlexGhiondea) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneSharedAccessSignatureTokenProvider.cs]: I don't understand this check -- are we trying to enforce that the resource is the first entry in the resource string?

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneSharedAccessSignatureTokenProvider.cs [hunk 3]
```diff
@@ -56,7 +56,11 @@ public TrackOneSharedAccessTokenProvider(SharedAccessSignature sharedAccessSigna
             Guard.ArgumentNotNullOrEmpty(nameof(resource), resource);
             Guard.ArgumentNotNegative(nameof(tokenValidityDuration), tokenValidityDuration);
 
-            if (!String.Equals(resource, SharedAccessSignature.Resource, StringComparison.InvariantCultureIgnoreCase))
+            // The requested resource may be a subpath of the current token resource, if a specific partition or other
+            // child object is referenced.  In that caes, the resource of the token should correspond to the starting
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneSharedAccessSignatureTokenProvider.cs]: nit: Typo on "case"

**Review by AlexGhiondea — APPROVED** [review]
Small question, but looks ok overall.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/ExponentialRetry.cs [hunk 4]
```diff
@@ -87,6 +87,20 @@ public sealed class ExponentialRetry : Retry
...
             new ExponentialRetry(_minimumBackoff, _maximumBackoff, _maximumRetryCount);
 
+        /// <summary>
+        ///   Allows internal access to the retry properties for compatibility shims.
+        /// </summary>
+        ///
+        /// <returns>The set of properties for the retry policy.</returns>
+        ///
+        /// <remarks>
+        ///     This method is intended to allow for compatibility shims to create the equivilent retry
```
Comment (jsquire) [inline sdk/eventhub/Azure.Messaging.EventHubs/src/ExponentialRetry.cs]: nit:  Fix spacing for `remarks`

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs (hunk 1 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs]: Will do;  that makes sense.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneSharedAccessSignatureTokenProvider.cs (hunk 2 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneSharedAccessSignatureTokenProvider.cs]: We're saying "If the resource that we're requesting a token for is the same as our current resource or is a child (subpath) of it, then we're good."      I'll punch up the comment a bit to try and clarify.

Token Resource:  `amqps://somehub.servicebus.com/hub-name`
Good: `amqps://somehub.servicebus.com/hub-name/partitions/0`
Bad: `amqps://somehub.servicebus.com/different-hub`
Bad: `amqps://different.servicebus.com/hub-name`

Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneSharedAccessSignatureTokenProvider.cs]: Rewrote and expanded the comment with examples.   I also overhauled the associated tests to be more aligned with the expected good/bad scenarios.

File: sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs (hunk 1 above)
Comment (jsquire) [reply sdk/eventhub/Azure.Messaging.EventHubs/src/Compatibility/TrackOneEventHubClient.cs]: Done, and fixed casing as well.

**Pull request merged**